### PR TITLE
`Programming exercises`: Improve feedback counter in result details

### PR DESCRIPTION
**Pull request by JohannesStoehr**
#### General

- [x] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [x] This is a small issue that I tested locally and was confirmed by another developer on a test server.
- [x] I chose a title conforming to the [naming conventions for pull requests](https://artemis-platform.readthedocs.io/en/latest/dev/guidelines/development-process.html#naming-conventions-for-github-pull-requests).
#### Client
- [x] I followed the [coding and design guidelines](https://docs.artemis.ase.in.tum.de/dev/guidelines/client/) and ensured that the layout is responsive.
- [x] I added multiple screenshots/screencasts of my UI changes.

### Motivation and Context

Before this PR, each result detail page would always show the number of tests, sca feedback and manual feedback in total, regardless whether the student clicked on the result on the top right to see all feedbacks or a specific task in the problem statement that only shows selected feedbacks for that task.

### Description

This PR fixes the behavior such that the number above the feedback list only reflects the feedbacks that are actually presented.

Additionally this PR localizes the ResultDetailsView in order to support German as well.

### Steps for Testing

Prerequisites:
- 1 Instructor
- 1 Student
- 1 Programming Exercise with SCA enabled and manual feedback

1. Log in to Artemis
2. Navigate to Course Administration
3. Participate in the exercise and pass some tests and get some code issues
4. Assess the exercise
5. Check that the result details always show the correct number of passed tests, sea feedback and manual feedback when clicking on the result on the top right as well as any task in the problem statement

#### Code Review
- [x] Review 1
- [x] Review 2
#### Manual Tests
- [ ] Test 1
- [ ] Test 2

### Screenshots

Showing all tests, which is wrong:
Bildschirm­foto 2022-11-03 um 18 44 51
Only showing the tests relevant for the current view:
Bildschirm­foto 2022-11-03 um 18 43 20

### CODE REVIEW

**Review by Strohgelaender — APPROVED** [review]
code looks good.

**Review by b-fein — DISMISSED** [review]
Code :+1:

File: src/main/webapp/i18n/de/result.json [hunk 1]
```diff
@@ -18,7 +18,24 @@
...
-                "title": "Bewertung"
+                "title": "Bewertung",
+                "feedback": "Feedback",
+                "codeIssue": {
+                    "name": "Codeissue",
+                    "title": "{{ scaCategory }} Issue in Datei {{ location }}",
+                    "line": "at line {{ line }}",
+                    "lines": "at lines {{ from }}-{{ to }}",
+                    "column": "column {{ column }}",
+                    "columns": "columns {{ from }}-{{ to }}"
```
Comment (b-fein) [inline src/main/webapp/i18n/de/result.json:29]: Copy-pasted from the English translation.

**Review by b-fein — CHANGES_REQUESTED** [review]

**Review by timo07k — DISMISSED** [review]
Correct number of passed tests and feedback is displayed for every view. Tested on TS1.

**Review by b-fein — DISMISSED** [review]
Code :+1:

File: src/main/webapp/app/exercises/shared/result/result-detail.component.ts [hunk 2]
```diff
@@ -364,11 +369,17 @@ export class ResultDetailComponent implements OnInit {
...
      */
-    private static getIssueLocation(issue: StaticCodeAnalysisIssue): string {
-        const lineText = !issue.endLine || issue.startLine === issue.endLine ? ` at line ${issue.startLine}` : ` at lines ${issue.startLine}-${issue.endLine}`;
+    private getIssueLocation(issue: StaticCodeAnalysisIssue): string {
+        const lineText =
+            ' ' + !issue.endLine || issue.startLine === issue.endLine
+                ? this.translateService.instant('artemisApp.result.detail.codeIssue.line', { line: issue.startLine })
+                : this.translateService.instant('artemisApp.result.detail.codeIssue.lines', { from: issue.startLine, to: issue.endLine });
         let columnText = '';
         if (issue.startColumn) {
```
Comment (julian-christl) [inline src/main/webapp/app/exercises/shared/result/result-detail.component.ts:378]:
```suggestion
```

File: src/main/webapp/app/exercises/shared/result/result-detail.component.ts [hunk 3]
```diff
@@ -364,11 +369,17 @@ export class ResultDetailComponent implements OnInit {
      * Builds the location string for a static code analysis issue
      * @param issue The sca issue
      */
-    private static getIssueLocation(issue: StaticCodeAnalysisIssue): string {
-        const lineText = !issue.endLine || issue.startLine === issue.endLine ? ` at line ${issue.startLine}` : ` at lines ${issue.startLine}-${issue.endLine}`;
+    private getIssueLocation(issue: StaticCodeAnalysisIssue): string {
+        const lineText =
+            ' ' + !issue.endLine || issue.startLine === issue.endLine
+                ? this.translateService.instant('artemisApp.result.detail.codeIssue.line', { line: issue.startLine })
+                : this.translateService.instant('artemisApp.result.detail.codeIssue.lines', { from: issue.startLine, to: issue.endLine });
```
Comment (julian-christl) [inline src/main/webapp/app/exercises/shared/result/result-detail.component.ts:376]:
```suggestion
        const lineText =
            !issue.endLine || issue.startLine === issue.endLine
                ? this.translateService.instant('artemisApp.result.detail.codeIssue.line', { line: issue.startLine })
                : this.translateService.instant('artemisApp.result.detail.codeIssue.lines', { from: issue.startLine, to: issue.endLine });
```

File: src/main/webapp/app/exercises/shared/result/result-detail.component.ts [hunk 4]
```diff
@@ -364,11 +369,17 @@ export class ResultDetailComponent implements OnInit {
...
         let columnText = '';
         if (issue.startColumn) {
-            columnText = !issue.endColumn || issue.startColumn === issue.endColumn ? ` column ${issue.startColumn}` : ` columns ${issue.startColumn}-${issue.endColumn}`;
+            columnText =
+                ' ' + !issue.endColumn || issue.startColumn === issue.endColumn
+                    ? this.translateService.instant('artemisApp.result.detail.codeIssue.column', { line: issue.startColumn })
+                    : this.translateService.instant('artemisApp.result.detail.codeIssue.columns', { from: issue.startColumn, to: issue.endColumn });
```
Comment (julian-christl) [inline src/main/webapp/app/exercises/shared/result/result-detail.component.ts:382]:
```suggestion
        const columnText =
            !issue.endColumn || issue.startColumn === issue.endColumn
                ? this.translateService.instant('artemisApp.result.detail.codeIssue.column', { line: issue.startColumn })
                : this.translateService.instant('artemisApp.result.detail.codeIssue.columns', { from: issue.startColumn, to: issue.endColumn });
```

File: src/main/webapp/app/exercises/shared/result/result-detail.component.ts [hunk 5]
```diff
@@ -364,11 +369,17 @@ export class ResultDetailComponent implements OnInit {
...
         }
         return issue.filePath + lineText + columnText;
```
Comment (julian-christl) [inline src/main/webapp/app/exercises/shared/result/result-detail.component.ts:384]:
```suggestion
        return issue.filePath + ' ' + lineText + (issue.startColumn ? ' ' + columnText : '');
```

File: src/main/webapp/app/exercises/shared/result/result-detail.component.ts [hunk 6]
```diff
@@ -475,13 +490,19 @@ export class ResultDetailComponent implements OnInit {
...
             if (positiveTestCasesWithoutDetailText.length > 0) {
+                this.numberOfAggregatedTestCases = positiveTestCasesWithoutDetailText.length;
                 return [
                     {
                         type: FeedbackItemType.Test,
-                        category: 'Feedback',
-                        title: positiveTestCasesWithoutDetailText.length + ' passed test' + (positiveTestCasesWithoutDetailText.length > 1 ? 's' : ''),
+                        category: this.showTestDetails
+                            ? this.translateService.instant('artemisApp.result.detail.test.name')
+                            : this.translateService.instant('artemisApp.result.detail.feedback'),
```
Comment (julian-christl) [inline src/main/webapp/app/exercises/shared/result/result-detail.component.ts:499]: Looking at those translations - would it make more sense to move this to the template? The component would need to be restructured then, though.

It's way out of scope for this PR, but I'd like to hear your opinion on that.

File: src/main/webapp/app/exercises/shared/result/result-detail.component.ts [hunk 7]
```diff
@@ -691,5 +713,23 @@ export class ResultDetailComponent implements OnInit {
...
+
+    private countFeedbacks() {
+        const testCaseList = this.filteredFeedbackList.filter((feedback) => feedback.type === FeedbackItemType.Test);
+        if (this.numberOfAggregatedTestCases && (this.showOnlyPositiveFeedback || !this.showOnlyNegativeFeedback)) {
+            this.testCaseCount = testCaseList.length + this.numberOfAggregatedTestCases - 1;
+            this.passedTestCaseCount = testCaseList.filter((feedback) => feedback.positive).length + this.numberOfAggregatedTestCases - 1;
+        } else {
+            this.testCaseCount = testCaseList.length;
+            this.passedTestCaseCount = testCaseList.filter((feedback) => feedback.positive).length;
+        }
```
Comment (julian-christl) [inline src/main/webapp/app/exercises/shared/result/result-detail.component.ts:728]: To me this is not intuitive why there is such a distinction. Could you please add some comments?

For example, why the `-1`? Both are counts so there should be no need for subtraction. If there is need for that, why?
Next: How do the two variables `this.showOnlyPositiveFeedback` and `this.showOnlyNegativeFeedback` play in the calculation. We filter for positive feedback either way and `this.numberOfAggregatedTestCases` is either 0 or contains a number of positive cases.

At least that's what I can see. So some educating documentation would be very much appreciated.

File: src/main/webapp/i18n/en/result.json [hunk 8]
```diff
@@ -18,7 +18,24 @@
                 }
             },
             "detail": {
-                "title": "Result"
+                "title": "Bewertung",
+                "feedback": "Feedback",
+                "codeIssue": {
+                    "name": "Code Issue",
+                    "title": "{{ scaCategory }} Issue in file {{ location }}",
```
Comment (julian-christl) [inline src/main/webapp/i18n/en/result.json:25]:
```suggestion
                    "title": "{{ scaCategory }} issue in file {{ location }}",
```

or

```suggestion
                    "title": "{{ scaCategory }} Issue in File {{ location }}",
```

**Review by julian-christl — COMMENTED** [review]
Take a look on what you want to change.

4 suggestions come from one place where I suggest restructuring that method. But as there are "deleted" lines, I can't do one single suggestion.

**Comment by julian-christl** [comment]
Another thing: Do you think it would make sense to mark that the view only shows partial results? Before the numbers were wrong but at least indicated that the result was only partial.

It could be as little as "(Partial)" in the title or a short text hinting at it.

Either way, good job. Especially with those translations. Was probably good fun ^^

File: src/main/webapp/app/exercises/shared/result/result-detail.component.ts (hunk 6 above)
Comment (JohannesStoehr) [reply src/main/webapp/app/exercises/shared/result/result-detail.component.ts:499]: I agree, a follow up might want to improve on this

**Review by julian-christl — APPROVED** [review]
Thanks for the improvements. Code approved :)

**Review by b-fein — APPROVED** [review]
Code re-approve.

**Review by kekskatze — APPROVED** [review]
Tested on testserver 2, everything worked as described

**Pull request merged**